### ISSUE TITLE
keras conv layer weight cannot be updated during distribution training mode

### ISSUE THREAD

**Issue by w19787** [tensorflow/tensorflow#37834]
OS: Ubuntu 18.04
Tensorflow: 2.1.0
Python: 3.7
GPUs: 4 gpus

trying to train a gan network on multiple GPUs. The network has a spectral normalization layer to update convolution layer weights. The sample code and error are provided as follows:

```python
import tensorflow as tf
strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
    conv = tf.keras.layers.Conv2D(5, 3, 1)
    x = tf.random.normal((1,5, 5, 3))
    conv(x)

@tf.function
def update_w(w, nw):
    w.assign(nw)

x2 = tf.random.normal((3,3, 3, 5))
strategy.experimental_run_v2(update_w, args=(conv.kernel, x2))
```
![image](https://user-images.githubusercontent.com/731496/77329724-5a234800-6d59-11ea-9ee4-a3b976178147.png)

**Comment by Saduf2019** [comment]
i have replicated the error faced on nightly, please find the [gist here](https://colab.sandbox.google.com/gist/Saduf2019/f9a25e7d853d7e41aebf3bfe473aaac0/37834.ipynb)

**Comment by anj-s** [comment]
@w19787 As the error message suggests we currently disallow aggregating a MirroredVariable in replica context since we aggregate the values before assigning (this is to ensure that the MirroredVariables remain in sync).

Would it be possible to work around this by updating the variable in cross replica context like the error message suggests?

**Comment by w19787** [comment]
i am unable to work around this issue in cross replica context with merge fn and extended.update. it can work on one gpu but still failed on multiple gpu.

when we create keras layer, i.e Conv2d, there is no argument to specify the aggregation parameter of weights, i think it should be the reason for this issue. I want to confirm whether the above understanding is correct, if so, why we design so? thanks.

**Comment by leosampaio** [comment]
A workaround for this is making a new variable creator and using `tf.variable_creator_scope`:
```
def variable_creator(next_creator, **kwargs):
    kwargs['aggregation'] = tf.VariableAggregation.ONLY_FIRST_REPLICA
    return next_creator(**kwargs)
with tf.variable_creator_scope(variable_creator):
    conv.build(input_shape)
```
Calling `build` inside your creator's scope will make the variables inside get the `aggregation` parameter. I chose `tf.VariableAggregation.ONLY_FIRST_REPLICA` but you may chose the `tf.VariableAggregation` that fits your use case better

**Comment by anj-s** [comment]
@w19787 Sorry for the late response. Recent changes have made it possible to call assign on mirrored variables in replica context(when there is no aggregation). Hopefully that should fix issues like this. However if you want to update your variable in cross replica context you can use something like:

```
strategy = tf.distribute.MirroredStrategy()
print("Devices used ", strategy.extended.worker_devices)
with strategy.scope():
    conv = tf.keras.layers.Conv2D(5, 3, 1)
    x = tf.random.normal((1,5, 5, 3))
    conv(x)

@tf.function
def update_w(w, nw):
    def merge_fn(strategy, w, nw):
      strategy.extended.update(w, lambda v,w:v.assign(w), args=(nw,))
    replica_ctx = tf.distribute.get_replica_context()
    replica_ctx.merge_call(merge_fn, args=(w, nw))

x2 = tf.random.normal((3,3, 3, 5))
print("before update ", conv.kernel[0])
strategy.experimental_run_v2(update_w, args=(conv.kernel, x2))
print("after update ", conv.kernel[0])
```

**Comment by w19787** [comment]
> @w19787 Sorry for the late response. Recent changes have made it possible to call assign on mirrored variables in replica context(when there is no aggregation). Hopefully that should fix issues like this. However if you want to update your variable in cross replica context you can use something like:
> 
> ```
> strategy = tf.distribute.MirroredStrategy()
> print("Devices used ", strategy.extended.worker_devices)
> with strategy.scope():
>     conv = tf.keras.layers.Conv2D(5, 3, 1)
>     x = tf.random.normal((1,5, 5, 3))
>     conv(x)
> 
> @tf.function
> def update_w(w, nw):
>     def merge_fn(strategy, w, nw):
>       strategy.extended.update(w, lambda v,w:v.assign(w), args=(nw,))
>     replica_ctx = tf.distribute.get_replica_context()
>     replica_ctx.merge_call(merge_fn, args=(w, nw))
> 
> x2 = tf.random.normal((3,3, 3, 5))
> print("before update ", conv.kernel[0])
> strategy.experimental_run_v2(update_w, args=(conv.kernel, x2))
> print("after update ", conv.kernel[0])
> ```

thanks a lot!
